lib/user/hash.js

```diff
@@ -0,0 +1,83 @@
+var crypto = require('crypto');
+var semver = require('semver');
+
+var SALT_LENGTH = 64;
+var HASH_ITERATIONS = 10000;
+var HASH_LENGTH = 256;
+var CIPHER = 'sha1';
+var SEPARATOR = ':';
+
+function genSalt() {
+  return crypto.randomBytes(SALT_LENGTH).toString('hex');
+}
+
+function pbkdf2(pwd, salt, iterations, length, cipher, cb) {
+  if (semver.satisfies(process.version, '>=0.12.0')) {
+    crypto.pbkdf2(pwd, salt, iterations, length, cipher, cb);
+  } else {
+    crypto.pbkdf2(pwd, salt, iterations, length, cb);
+  }
+}
+
+/**
+ * @callback saltAndHashCallback
+ * @param {Error}
+ * @param {String} responseMessage
+ */
+
+/**
+ * Generates a random salt and encrypts the given password
+ * @param  {String}   pwd Password to be encrypted
+ * @param  {saltAndHashCallback} cb Node-style callback
+ */
+exports.saltAndHash = function(pwd, cb) {
+  var salt = genSalt();
+  pbkdf2(pwd, salt, HASH_ITERATIONS, HASH_LENGTH, CIPHER, function(err, hashed) {
+    if (err) {
+      return cb(err);
+    }
+
+    var finalHash = [CIPHER, HASH_ITERATIONS, hashed.toString('hex'), salt].join(SEPARATOR);
+    cb(null, finalHash);
+  });
+};
+
+/**
+ * @callback verifyCallback
```

Comment:
this is jsdoc' way of specifiying typing for callback functions [jsdocs docs](http://usejsdoc.org/tags-callback.html). This one is used at line 55 [here](https://github.com/feedhenry-raincatcher/raincatcher-user/pull/19/files/484c0cf18bdd43bf1f49363343646a0e772c0864#diff-c00fbd7311ea36536a97f3ca438d9fd2R55).